clients/src/test/java/org/apache/kafka/common/requests/StopReplicaRequestTest.java

```diff
@@ -44,22 +52,144 @@ public void testUnsupportedVersion() {
+            }
+        }
+
         for (short version = STOP_REPLICA.oldestVersion(); version < STOP_REPLICA.latestVersion(); version++) {
             StopReplicaRequest.Builder builder = new StopReplicaRequest.Builder(version,
-                    0, 0, 0L, false, Collections.emptyList());
+                    0, 0, 0L, false, topicStates);
             StopReplicaRequest request = builder.build();
             StopReplicaResponse response = request.getErrorResponse(0,
                     new ClusterAuthorizationException("Not authorized"));
             assertEquals(Errors.CLUSTER_AUTHORIZATION_FAILED, response.error());
+            assertEquals(expectedPartitions, new HashSet<>(response.partitionErrors()));
+        }
+    }
+
+    @Test
+    public void testBuilderNormalization() {
+        List<StopReplicaTopicState> topicStates = topicStates();
+
+        Set<TopicPartition> expectedPartitions = new HashSet<>();
+        for (StopReplicaTopicState topicState : topicStates) {
+            for (StopReplicaPartitionState partitionState: topicState.partitionStates()) {
+                expectedPartitions.add(
+                    new TopicPartition(topicState.topicName(), partitionState.partitionIndex()));
+            }
+        }
+
+        Map<TopicPartition, StopReplicaPartitionState> expectedPartitionStates = new HashMap<>();
+        for (StopReplicaTopicState topicState : topicStates) {
+            for (StopReplicaPartitionState partitionState: topicState.partitionStates()) {
+                expectedPartitionStates.put(
+                    new TopicPartition(topicState.topicName(), partitionState.partitionIndex()),
+                    partitionState);
+            }
+        }
+
+        for (short version = STOP_REPLICA.oldestVersion(); version < STOP_REPLICA.latestVersion(); version++) {
+            StopReplicaRequest request = new StopReplicaRequest.Builder(version, 0, 1, 0,
+                true, topicStates).build(version);
+            StopReplicaRequestData data = request.data();
+
+            if (version < 1) {
+                Set<TopicPartition> partitions = new HashSet<>();
+                for (StopReplicaPartitionV0 partition : data.ungroupedPartitions()) {
+                    partitions.add(new TopicPartition(partition.topicName(), partition.partitionIndex()));
+                }
+                assertEquals(expectedPartitions, partitions);
+            } else if (version < 3) {
+                Set<TopicPartition> partitions = new HashSet<>();
+                for (StopReplicaTopicV1 topic : data.topics()) {
+                    for (Integer partition : topic.partitionIndexes()) {
+                        partitions.add(new TopicPartition(topic.name(), partition));
+                    }
+                }
+                assertEquals(expectedPartitions, partitions);
+            } else {
+                Map<TopicPartition, StopReplicaPartitionState> partitionStates = new HashMap<>();
```

Comment:
I have considered this as well but I haven't done it because it is only used in tests so far. I think that the downside is that using this in core is not optimal thus I am a bit reluctant to provide it. I mean, allocating and populating the `Map` is not necessary, especially when the controller and the brokers use the latest version of the API.